integration-tests/src/integrationTest/java/pl/allegro/tech/hermes/integrationtests/KafkaRetransmissionServiceTest.java

```diff
@@ -0,0 +1,188 @@
+import com.google.common.collect.Multimaps;
+import com.jayway.awaitility.Duration;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.springframework.test.web.reactive.server.WebTestClient;
+import pl.allegro.tech.hermes.api.ContentType;
+import pl.allegro.tech.hermes.api.OffsetRetransmissionDate;
+import pl.allegro.tech.hermes.api.PatchData;
+import pl.allegro.tech.hermes.api.Subscription;
+import pl.allegro.tech.hermes.api.Topic;
+import pl.allegro.tech.hermes.common.kafka.offset.PartitionOffset;
+import pl.allegro.tech.hermes.consumers.supervisor.process.RunningSubscriptionStatus;
+import pl.allegro.tech.hermes.integrationtests.setup.HermesExtension;
+import pl.allegro.tech.hermes.integrationtests.subscriber.TestSubscriber;
+import pl.allegro.tech.hermes.integrationtests.subscriber.TestSubscribersExtension;
+import pl.allegro.tech.hermes.management.infrastructure.kafka.MultiDCOffsetChangeSummary;
+import pl.allegro.tech.hermes.test.helper.avro.AvroUser;
+import pl.allegro.tech.hermes.test.helper.message.TestMessage;
+
+import java.time.Clock;
+import java.time.OffsetDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Predicate;
+
+import static com.jayway.awaitility.Awaitility.waitAtMost;
+import static java.util.stream.IntStream.range;
+import static org.assertj.core.api.Assertions.assertThat;
+import static pl.allegro.tech.hermes.api.PatchData.patchData;
+import static pl.allegro.tech.hermes.consumers.supervisor.process.Signal.SignalType.COMMIT;
+import static pl.allegro.tech.hermes.consumers.supervisor.process.Signal.SignalType.UPDATE_TOPIC;
+import static pl.allegro.tech.hermes.test.helper.builder.SubscriptionBuilder.subscriptionWithRandomName;
+import static pl.allegro.tech.hermes.test.helper.builder.TopicBuilder.topicWithRandomName;
+import static pl.allegro.tech.hermes.test.helper.endpoint.TimeoutAdjuster.adjust;
+
+
+// TODO potentially should be moved to slow integrationTest
+public class KafkaRetransmissionServiceTest {
+
+    private static final String PRIMARY_KAFKA_CLUSTER_NAME = "primary-dc";
+
+    private final AvroUser avroUser = new AvroUser();
+
+    private final Clock clock = Clock.systemDefaultZone();
+
+    private final List<String> messages = new ArrayList<>() {{
+        range(0, 4).forEach(i -> add(TestMessage.random().body()));
+    }};
+
+    private final List<String> messages2 = new ArrayList<>() {{
+        range(5, 6).forEach(i -> add(TestMessage.random().body()));
+    }};
+
+    @RegisterExtension
+    public static final HermesExtension hermes = new HermesExtension();
+
+    @RegisterExtension
+    public static final TestSubscribersExtension subscribers = new TestSubscribersExtension();
+
+    @Test
+    public void shouldMoveOffsetNearGivenTimestamp() throws InterruptedException {
+        // given
+        TestSubscriber subscriber = subscribers.createSubscriber();
```

Comment:
**nit:** these "given" sections are quite long, what do you think about separating the different parts with empty lines to improve readability?

I wonder what you think about wrapping these two lines of code in some private method (just to make given section a bit smaller)?

```java
publishMessagesAndWaitUntilReceived(String topicName, List<String> messages) { // Perhaps not the best name for this method
   messages.forEach(message -> hermes.api().publish(topicName, message));
   messages.forEach(subscriber::waitUntilReceived);
}
```

But both comments are probably a matter of preference so it is up to you 😄